`__ray_kill__` does not work as expected

Related PR: https://github.com/ray-project/ray/pull/6523

## Analysis

In the test case `test_kill`, the specified timeout value of `ray.wait` is too small, so it's very likely that the actor is not created yet when reached the timeout. So the remote call to `hang` is still pending to be pushed and the assertion `assert len(ready) == 0` can be passed.

https://github.com/ray-project/ray/blob/0a5d0109a406dedb389e0c71fcc9129e45ae7b92/python/ray/tests/test_actor.py#L1423-L1429

Then the `__ray_kill__` operation is initiated, the kill operation is first cached at local. After the actor is alive, the kill operation is pushed to the remote actor prior to the previous remote task, because in `CoreWorkerDirectActorTaskSubmitter::SendPendingTasks`, the `KillActor` messages are always sent before any actor tasks.

https://github.com/ray-project/ray/blob/0a5d0109a406dedb389e0c71fcc9129e45ae7b92/src/ray/core_worker/transport/direct_actor_transport.cc#L125-L144

At the remote side, the worker receives the `KillActor` message first and just invokes the `CoreWorker::Shutdown`, and then both the `io_service_` and the `task_execution_service_` are stopped. The `hang` function never get the chance to run.

https://github.com/ray-project/ray/blob/0a5d0109a406dedb389e0c71fcc9129e45ae7b92/src/ray/core_worker/core_worker.cc#L1156-L1157

https://github.com/ray-project/ray/blob/0a5d0109a406dedb389e0c71fcc9129e45ae7b92/src/ray/core_worker/core_worker.cc#L259-L264

Eventually, **the test case will pass but it can't verify that the `__ray_kill__` works.**

## Test case update

If we change the timeout value to `5`, the test case will fail with below message. That's because `ray.get(ray.ObjectID.from_random())` actually throws an `UnreconstructableError`, not `Never returns` as the comment says.

```python
>       assert len(ready) == 0
E       assert 1 == 0
E         -1
E         +0

python/ray/tests/test_actor.py:1426: AssertionError
```

Then we change the implementation of `hang` to an infinite loop with `time.sleep(1)`, now the test case hangs forever as I expected, but the worker still won't quit. I think it's because the `hang` function is still running and `task_execution_service_.stop()` won't abort the execution of `hang`.

Final diff:

```diff
diff --git a/python/ray/tests/test_actor.py b/python/ray/tests/test_actor.py
index 9c43c65fc..98956b297 100644
--- a/python/ray/tests/test_actor.py
+++ b/python/ray/tests/test_actor.py
@@ -1418,11 +1418,14 @@ def test_kill(ray_start_regular):
     class Actor:
         def hang(self):
             # Never returns.
-            ray.get(ray.ObjectID.from_random())
+            while True:
+                time.sleep(1)
 
     actor = Actor.remote()
     result = actor.hang.remote()
-    ready, _ = ray.wait([result], timeout=0.1)
+    ready, _ = ray.wait([result], timeout=5)
+    if len(ready) > 0:
+        print(ray.get(result))
     assert len(ready) == 0
     actor.__ray_kill__()
     with pytest.raises(ray.exceptions.RayActorError):
```

--- a/python/ray/tests/test_actor.py
+++ b/python/ray/tests/test_actor.py
@@ -1413,12 +1413,12 @@ def test_kill(ray_start_regular):
     @ray.remote
     class Actor:
         def hang(self):
-            # Never returns.
-            ray.get(ray.ObjectID.from_random())
+            while True:
+                time.sleep(1)
 
     actor = Actor.remote()
     result = actor.hang.remote()
-    ready, _ = ray.wait([result], timeout=0.1)
+    ready, _ = ray.wait([result], timeout=0.5)
     assert len(ready) == 0
     actor.__ray_kill__()
     with pytest.raises(ray.exceptions.RayActorError):

--- a/src/ray/core_worker/core_worker.cc
+++ b/src/ray/core_worker/core_worker.cc
@@ -1153,8 +1153,11 @@ void CoreWorker::HandleKillActor(const rpc::KillActorRequest &request,
     send_reply_callback(Status::Invalid(msg), nullptr, nullptr);
     return;
   }
-  RAY_LOG(INFO) << "Got KillActor, shutting down...";
-  Shutdown();
+  RAY_LOG(INFO) << "Got KillActor, exiting immediately...";
+  if (log_dir_ != "") {
+    RayLog::ShutDownRayLog();
+  }
+  exit(1);
 }
 
 void CoreWorker::HandleGetCoreWorkerStats(const rpc::GetCoreWorkerStatsRequest &request,

